transports/websys/src/websocket.rs

```diff
@@ -0,0 +1,321 @@
+            format!("[{ip}]:{port}")
+        }
+        (Some(Protocol::Dns(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dns4(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dns6(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dnsaddr(h)), Some(Protocol::Tcp(port))) => {
+            format!("{}:{}", &h, port)
+        }
+        _ => return None,
+    };
+
+    let (scheme, wspath) = match protocols.next() {
+        Some(Protocol::Ws(path)) => ("ws", path.into_owned()),
+        Some(Protocol::Wss(path)) => ("wss", path.into_owned()),
+        _ => return None,
+    };
+
+    // TODO: handle PeerId
+    Some(format!("{scheme}://{host_port}{wspath}"))
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("js function error {0}")]
+    JsError(String),
+    #[error("operation not supported")]
+    NotSupported,
+}
+
+/// A Websocket connection created by the [`Transport`].
+pub struct Connection {
+    /// We need to use Mutex as libp2p requires this to be Send.
+    shared: Arc<Mutex<Shared>>,
+}
+
+struct Shared {
+    opened: bool,
+    closed: bool,
+    error: bool,
+    data: VecDeque<u8>,
+    waker: Option<Waker>,
+    socket: SendWrapper<WebSocket>,
+    closures: Option<SendWrapper<Closures>>,
+}
+
+type Closures = (
+    Closure<dyn FnMut()>,
+    Closure<dyn FnMut(MessageEvent)>,
+    Closure<dyn FnMut(web_sys::Event)>,
+    Closure<dyn FnMut(web_sys::CloseEvent)>,
+);
+
+impl Connection {
+    fn new(socket: WebSocket) -> Self {
+        socket.set_binary_type(web_sys::BinaryType::Arraybuffer);
+
+        let shared = Arc::new(Mutex::new(Shared {
+            opened: false,
+            closed: false,
+            error: false,
+            data: VecDeque::with_capacity(1 << 16),
+            waker: None,
+            socket: SendWrapper::new(socket.clone()),
+            closures: None,
+        }));
+
+        let open_callback = Closure::<dyn FnMut()>::new({
+            let shared = shared.clone();
+            move || {
+                let mut locked = shared.lock();
+                locked.opened = true;
+                if let Some(waker) = &locked.waker {
+                    waker.wake_by_ref();
+                }
+            }
+        });
+        socket.set_onopen(Some(open_callback.as_ref().unchecked_ref()));
+
+        let message_callback = Closure::<dyn FnMut(_)>::new({
+            let shared = shared.clone();
+            move |e: MessageEvent| {
+                if let Ok(abuf) = e.data().dyn_into::<js_sys::ArrayBuffer>() {
+                    let mut locked = shared.lock();
+                    let bytes = js_sys::Uint8Array::new(&abuf).to_vec();
+                    locked.data.extend(bytes.into_iter());
+                    if let Some(waker) = &locked.waker {
+                        waker.wake_by_ref();
+                    }
+                } else {
+                    panic!("Unexpected data format {:?}", e.data());
```

Comment:
Is this definitely unreachable? Can we make this a `debug_assert` instead?

Reply:
From what I understand libp2p Websocket doesn't send blobs (web_sys::Blob), but I can change it if you think is more appropriate.

Reply:
Yeah, I'd rather not have `panic!`s that are technically reachable, esp. not in parts of the code that are handling network input.